### ISSUE TITLE
More simplifications for List.take and List.drop

### ISSUE THREAD

**Issue by morteako** [jfmengels/elm-review-simplify#268]
# More simplifications for List.take and List.drop

Some suggestions for List.take and List.drop simplifications.
I think take and map are functions that are used a lot, so new rules would probably be applied in many codebases 🚀  

(`n` is refering to the Int argument , as in `List.take n list`)

### List.drop with negative n
```elm
List.drop -1 list
--> list
```
### List.length on List.take with non-negative n
```elm
List.length (List.take 10 list)
--> 10
```

### Literal lists with size <= n
```elm
List.take 5 [a,b,c]
--> [a,b,c]

List.drop 5 [a,b,c]
--> []

List.take 1 (List.singleton x)
--> List.singleton x

List.drop 1 (List.singleton x)
--> []
```

### Take on take, drop on drop
```elm
list
|> List.take n
|> List.take m
--> list |> List.take (min n m)

list
|> List.drop n
|> List.drop m
--> list |> List.drop (n + m)
```

### List.repeat with non-negative n <= repeat arg
```elm
List.take 3 (List.repeat 5 x )
--> List.repeat 2 x

List.drop 3 (List.repeat 5 x )
--> List.drop 2 x
```

### List.range with size >= n
```elm
List.take 3 (List.range 0 10)
--> List.range 0 2

List.drop 3 (List.range 0 10)
--> List.range 3 10
```

### Optimizations when combining with List.map/filterMap
```elm
-- Same for List.drop and same for List.filterMap
list
|> List.map f
|> List.take n
--> 
list
|> List.take n
|> List.map f
```

---
### Generalization question

Some of these can be generalized more, f.ex.

```elm
-- n > 0
List.drop n (List.repeat m x )
--> List.repeat (m-n) x
```
But I am not sure about the clarity of the resulting code and also about how it will impact further simplifications.
So I went with the simpler one. Open for suggestions

---
Feedback and tips welcome! And of course more suggestions and corrections.

I can probably look into implementing most of these myself.

**Comment by lue-bird** [comment]
Nice.
- ```elm
  List.length (List.take 10 list)
  --> 10
  ```
  doesn't work because list can have less than 10 elements.

- The take and drop ones are already tracked in https://github.com/jfmengels/elm-review-simplify/issues/2

- ```elm
  List.take 3 (List.repeat 5 x)
  --> List.repeat 2 x
  ```
  should be repeat 3

Other than those nitpicks, looks good

**Comment by morteako** [comment]
> * ```elm
>   List.length (List.take 10 list)
>   --> 10
>   ```
>   doesn't work because list can have less than 10 elements.

Yes, of course. My bad. This was a bit of a hasty late suggestion. Removed

> * The take and drop ones are already tracked in [Simplifications to add #2](https://github.com/jfmengels/elm-review-simplify/issues/2)
> * The take and drop after map are already tracked in [Simplify to more optimized mapping operations #112](https://github.com/jfmengels/elm-review-simplify/issues/112)
Ah. Makes sense. I removed them from this issue 

> * ```elm
>   List.take 3 (List.repeat 5 x)
>   --> List.repeat 2 x
>   ```
>   should be repeat 3

Fixed! Thanks

**Comment by jfmengels** [comment]
Nice suggestions :+1: 

Other things we could do:

```elm
List.drop 2 (x :: y :: z :: list)
--> z :: list

List.drop 2 (x :: list)
--> List.drop 1 list

List.take 2 (x :: list)
--> x :: List.take 1 list
```

---

@morteako Would you like to work on these simplifications? Asking before @lue-bird snatches them away :sweat_smile:

**Comment by morteako** [comment]
@jfmengels 
I will at least take a look at the first one , and try to include
```elm
List.drop 2 ([x,y,z] ++ list)
--> z :: list
```


I am not personally sure about
```elm
List.take 2 (x :: list)
--> x :: List.take 1 list
```
It should be better for performance, but i think maybe the resulting code is a bit more unclear.
